### ISSUE TITLE
Reference message types from parent package

### ISSUE THREAD

**Issue by mmmries** [tomas-abrahamsson/gpb#103]
I have several protobuf files that reference a field from a parent package in their definition like this:

```atlas.proto
package atlas;

message Error {
  optional string fields = 1;
  repeated string messages = 2;
}
```

```atlas/abacus/account.proto
package atlas.abacus;

message Account {
  extensions 200 to max;
  optional string guid = 1;
  extend Account {
    repeated Error errors = 200;
  }
}
```

The protobuf compiler that google publishes resolves the `repeated Error` back to the `atlas.Error` message type. I tried to take a stab at supporting this myself in `gpb`, but wasn't sure of the best way to proceed since it looks like field refs are resolved at a point in the code where we don't know which message the field is being defined in, or which packages I should consider to look for the reference.

Is this something you are interested in supporting? Should I handle it similar to how you handle the extension refs? It looks like you do a 2-pass approach for those where you generate a list of candidates on the first pass and then pick one during the second pass.

**Comment by tomas-abrahamsson** [comment]
Yes to supporting it. The aim has always been that it should allow the same input as google's protobuf. I think the 2-pass approach was to be able to extend messages or enums in other packages.

I think the reason for the failure is simply the order in which things happens when post-processing, see [`gpb_parse:post_process_all_files/2`](https://github.com/tomas-abrahamsson/gpb/blob/3.26.7/src/gpb_parse.yrl#L409L417): First, names are resolved, then `extends` are handled. But in this case, there is a need to resolve _after_ handling `extends` too, so that `Error` gets resolved.

At the moment, I'm not exactly sure where to best put the call to do a repeated resolve. Need to ponder a bit.

**Comment by mmmries** [comment]
What if we did the following sequence:

* resolve all message names
* resolve all enum names
* resolve all extensions
* resolve all field names + type refs

Fields always have to resolve their type to one of the built-in scalar types or an enum or a message right? So as long as we've resolved all message and enum names first I think we can still do all fields + extensions in a single pass.

**Comment by tomas-abrahamsson** [comment]
Yes, that could work. It is true that fields are either built-in types, or refer to enums or messages.

**Comment by tomas-abrahamsson** [comment]
I think I found the bug, I was wrong earlier guessing about it. Would you be able to try out the branch `tmp-extend-with-msg-in-enclosing-pkg` to see if this does the trick for you? If so, I'll make a new release.

**Comment by mmmries** [comment]
@tomas-abrahamsson Nice! That branch compiles almost my entire library of proto defintions (132 files, 499 message definitions). The only case that it doesn't appear to cover is a case that looks like this:

```protobuf a/foo.proto
package a;

message Foo {
  extensions 200 to max;
  optional string id = 1;
}
```

```protobuf b/bar.proto
package b;

message Bar {
  optional string id = 1;
}

extend a.Foo {
  optional Bar b = 200;
}  
```

It appears that the protobuf compiler is following the package ancestors for both the scope of the message being extended as well as the scope of the file in which it is defined. I can probably open that as a separate issue though because it's very easy for me to workaround.

**Comment by tomas-abrahamsson** [comment]
@mmmries Thanks for trying it, good to know it was a step forward. No need from my side for a new issue, I'll look into it.

**Comment by tomas-abrahamsson** [comment]
I've just (force) pushed a new version of the branch, still `tmp-extend-with-msg-in-enclosing-pkg`, which I  hope solves both problems. Feel free to give it a spin!

**Comment by mmmries** [comment]
Pulled in the update and sure enough it compiled my entire proto library with no modifications at all! Thanks for quick response on this. I'll try to keep contributing back the project wherever I can.

**Comment by tomas-abrahamsson** [comment]
Super! I've merged it to master and pushed it as 3.26.8, so I'll close this issue. (I'll be remove the temporary branch shortly.) Thanks for the useful reproducers! I included them in the unit tests to make sure things continue to work in the future too.
